File: kreuzberg/packages/java/src/main/java/dev/kreuzberg/HierarchicalBlock.java
package dev.kreuzberg;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * A text block with hierarchy level assignment.
 *
 * <p>
 * Represents a block of text with semantic heading information extracted from
 * font size clustering and hierarchical analysis.
 *
 * @param text
 *            the text content of this block
 * @param fontSize
 *            the font size of the text
 * @param level
 *            the hierarchy level (h1-h6 or body)
 * @param bbox
 *            bounding box as [left, top, right, bottom] in PDF units, or null
 */
public record HierarchicalBlock(@JsonProperty("text") String text, @JsonProperty("font_size") float fontSize,
		@JsonProperty("level") String level, @JsonProperty("bbox") float[] bbox) {
	@JsonCreator
	public HierarchicalBlock(@JsonProperty("text") String text, @JsonProperty("font_size") float fontSize,
			@JsonProperty("level") String level, @JsonProperty("bbox") float[] bbox) {
		this.text = text != null ? text : "";
		this.fontSize = fontSize;
		this.level = level != null ? level : "body";
		this.bbox = bbox;
	}
}

File: kreuzberg/packages/php/src/Types/HierarchicalBlock.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * A text block with hierarchy level assignment.
 *
 * @property-read string $text The text content of this block
 * @property-read float $fontSize The font size of the text
 * @property-read string $level The hierarchy level (h1-h6 or body)
 * @property-read ?array<float> $bbox Bounding box (left, top, right, bottom)
 */
readonly class HierarchicalBlock
{
    /**
     * @param ?array<float> $bbox
     */
    public function __construct(
        public string $text,
        public float $fontSize,
        public string $level,
        public ?array $bbox = null,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $text = isset($data['text']) && is_string($data['text']) ? $data['text'] : '';
        $fontSize = isset($data['font_size']) ? (is_float($data['font_size']) || is_int($data['font_size']) ? (float) $data['font_size'] : 0.0) : 0.0;
        $level = isset($data['level']) && is_string($data['level']) ? $data['level'] : 'body';
        $bboxRaw = $data['bbox'] ?? null;

        /** @var ?array<float> $bbox */
        $bbox = is_array($bboxRaw) ? array_map(static fn (mixed $v): float => is_float($v) || is_int($v) ? (float) $v : 0.0, $bboxRaw) : null;

        return new self(
            text: $text,
            fontSize: $fontSize,
            level: $level,
            bbox: $bbox,
        );
    }
}
